File: src/services/openMeteoService.ts
/**
 * Open-Meteo Weather API Integration Service
 * Fetches real-time ambient climate telemetry and 16-day rainfall projections.
 */

export interface WeatherData {
  success: boolean;
  source: 'open-meteo' | 'simulated-meteorological';
  temperature: number; // °C
  humidity: number; // %
  rainfallCurrent: number; // mm
  rainfallForecast16d: number; // mm
  annualizedRainfallEst: number; // mm
  windSpeed: number; // km/h
  weatherCode: number;
  daily?: {
    precipitation_sum?: number[];
    temperature_2m_max?: number[];
    temperature_2m_min?: number[];
  };
}

export class OpenMeteoService {
  private baseUrl: string;

  constructor(baseUrl: string = 'https://api.open-meteo.com/v1/forecast') {
    this.baseUrl = baseUrl;
  }

  /**
   * Fetch real-time weather and 16-day rainfall projections for given GPS coordinates.
   */
  async getForecast(lat: number, lon: number): Promise<WeatherData> {
    const url = `${this.baseUrl}?latitude=${lat}&longitude=${lon}&current=temperature_2m,relative_humidity_2m,precipitation,rain,weather_code,wind_speed_10m&daily=precipitation_sum,temperature_2m_max,temperature_2m_min&forecast_days=16&timezone=auto`;

    try {
      const response = await fetch(url, {
        headers: { Accept: 'application/json' },
      });

      if (!response.ok) {
        throw new Error(`Open-Meteo responded with status ${response.status}`);
      }

      const data = await response.json();
      const current = data.current || {};
      const daily = data.daily || {};
      const totalRainfall16Days = (daily.precipitation_sum || []).reduce(
        (acc: number, val: number) => acc + (val || 0),
        0
      );

      return {
        success: true,
        source: 'open-meteo',
        temperature: current.temperature_2m ?? 28.4,
        humidity: current.relative_humidity_2m ?? 73,
        rainfallCurrent: current.precipitation ?? 0,
        rainfallForecast16d: Math.round(totalRainfall16Days * 10) / 10,
        annualizedRainfallEst: Math.round(totalRainfall16Days * 22 + 400),
        windSpeed: current.wind_speed_10m ?? 12.5,
        weatherCode: current.weather_code ?? 1,
        daily,
      };
    } catch (error) {
      console.warn('OpenMeteoService falling back to regional meteorological model:', error);
      return this.getFallbackWeather(lat);
    }
  }

  /**
   * Regional simulated baseline when offline or rate-limited.
   */
  private getFallbackWeather(lat: number): WeatherData {
    const isTropical = Math.abs(lat) < 23.5;
    return {
      success: true,
      source: 'simulated-meteorological',
      temperature: isTropical ? 28.5 : 22.0,
      humidity: isTropical ? 72 : 58,
      rainfallCurrent: 1.5,
      rainfallForecast16d: 44.0,
      annualizedRainfallEst: 840,
      windSpeed: 11.5,
      weatherCode: 2,
      daily: {
        precipitation_sum: [2, 5, 0, 12, 18, 4, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0],
      },
    };
  }
}

export const openMeteoService = new OpenMeteoService();

File: backend/services/open_meteo_service.py
import json
import urllib.request
import urllib.parse
import asyncio
from typing import Dict, Any
from backend.config import settings

class OpenMeteoService:
    def __init__(self, base_url: str = settings.OPEN_METEO_API_URL):
        self.base_url = base_url

    def _sync_fetch(self, url: str) -> Dict[str, Any]:
        req = urllib.request.Request(url, headers={"User-Agent": "AgroXAI-Service/1.0", "Accept": "application/json"})
        with urllib.request.urlopen(req, timeout=6.0) as resp:
            return json.loads(resp.read().decode("utf-8"))

    async def get_weather(self, lat: float, lon: float) -> Dict[str, Any]:
        """
        Fetch real-time ambient climate and 16-day forecast using Open-Meteo REST API.
        """
        params = {
            "latitude": lat,
            "longitude": lon,
            "current": "temperature_2m,relative_humidity_2m,precipitation,rain,weather_code,wind_speed_10m",
            "daily": "precipitation_sum,temperature_2m_max,temperature_2m_min",
            "forecast_days": 16,
            "timezone": "auto",
        }
        query_string = urllib.parse.urlencode(params)
        url = f"{self.base_url}?{query_string}"
        
        try:
            # Run blocking urllib request in threadpool
            data = await asyncio.to_thread(self._sync_fetch, url)
            
            current = data.get("current", {})
            daily = data.get("daily", {})
            precip_list = daily.get("precipitation_sum", [])
            forecast_16d = round(sum(p for p in precip_list if p is not None), 1)
            
            return {
                "success": True,
                "source": "open-meteo",
                "temperature": current.get("temperature_2m", 28.4),
                "humidity": current.get("relative_humidity_2m", 73),
                "rainfallCurrent": current.get("precipitation", 0.0),
                "rainfallForecast16d": forecast_16d,
                "annualizedRainfallEst": round(forecast_16d * 22 + 400),
                "windSpeed": current.get("wind_speed_10m", 12.5),
                "weatherCode": current.get("weather_code", 1),
                "daily": daily,
            }
        except Exception as e:
            # Fallback deterministic meteorological estimate
            return {
                "success": True,
                "source": "simulated-meteorological",
                "temperature": 28.5,
                "humidity": 72.0,
                "rainfallCurrent": 1.5,
                "rainfallForecast16d": 44.0,
                "annualizedRainfallEst": 840,
                "windSpeed": 11.5,
                "weatherCode": 2,
                "error": str(e),
            }

open_meteo_service = OpenMeteoService()
